%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Do not modify this file since it was automatically generated from:
% 
%  999.NonDocumentedObjects.R
% 
% by the Rdoc compiler part of the R.oo package.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\name{Non-documented objects}
\alias{Non-documented objects}
\title{Non-documented objects}


% Other missing docs
\alias{as.character.MultiSourceCopyNumberNormalization}
\alias{extractTupleOfDataFiles}
\alias{fitOne}
\alias{getAllNames}
\alias{getAsteriskTags.TotalCnSmoothing}
\alias{getFitAromaUgpFile}
\alias{getInputDataSets}
\alias{getOutputDataSets}
\alias{getOutputDataSets.MultiSourceCopyNumberNormalization}
\alias{getOutputFiles.TotalCnSmoothing}
\alias{getOutputPaths}
\alias{getOutputPaths.MultiSourceCopyNumberNormalization}
\alias{getParameters.MultiSourceCopyNumberNormalization}
\alias{getParameters.TotalCnSmoothing}
\alias{getParametersAsString.MultiSourceCopyNumberNormalization}
\alias{getPath.TotalCnSmoothing}
\alias{getRootPath.TotalCnSmoothing}
\alias{getSmoothedDataSets}
\alias{getSubsetToFit}
\alias{getTargetPositions}
\alias{getTargetPositions.TotalCnSmoothing}
\alias{getTargetUgpFile}
\alias{getTargetUgpFile.TotalCnSmoothing}
\alias{makeSmoothSplinePredict}
\alias{makeSmoothSplinePredict.numeric}
\alias{nbrOfDataSets}
\alias{nbrOfDataSets.MultiSourceCopyNumberNormalization}
\alias{normalizeOne}
\alias{normalizeOne.MultiSourceCopyNumberNormalization}
\alias{normalizePrincipalCurve}
\alias{pairs2}
\alias{pairs2.principal.curve}
\alias{process.TotalCnSmoothing}
\alias{getAsteriskTags.TotalCnKernelSmoothing}
\alias{getParameters.TotalCnKernelSmoothing}
\alias{smoothRawCopyNumbers}
\alias{smoothRawCopyNumbers.TotalCnKernelSmoothing}
\alias{smoothRawCopyNumbers.TotalCnSmoothing}
\alias{asCartesianSnpData}
\alias{asCartesianSnpData.CartesianSnpData}
\alias{asCartesianSnpData.PolarSnpData}
\alias{asCartesianSnpData.TotalFracBSnpData}
\alias{asPolarSnpData}
\alias{asPolarSnpData.CartesianSnpData}
\alias{asPolarSnpData.PolarSnpData}
\alias{asPolarSnpData.TotalFracBSnpData}
\alias{asTotalFracBSnpData}
\alias{asTotalFracBSnpData.CartesianSnpData}
\alias{asTotalFracBSnpData.PolarSnpData}
\alias{asTotalFracBSnpData.TotalFracBSnpData}
\alias{callGenotypes}
\alias{callGenotypes.SnpData}
\alias{callGenotypes.TotalFracBSnpData}
\alias{CartesianSnpData}
\alias{pairedBoost}
\alias{pairedBoost.CartesianSnpData}
\alias{pairedBoost.PolarSnpData}
\alias{pairedBoost.TotalFracBSnpData}
\alias{plot.CartesianSnpData}
\alias{plot.PolarSnpData}
\alias{plot.SnpData}
\alias{plot.TotalFracBSnpData}
\alias{PolarSnpData}
\alias{SnpData}
\alias{TotalFracBSnpData}
\alias{as.character.TumorBoostNormalization}
\alias{getDataSets}
\alias{getDataSets.TumorBoostNormalization}
\alias{getInputDataSet.TumorBoostNormalization}
\alias{getNormalDataSet}
\alias{getNormalDataSet.TumorBoostNormalization}
\alias{getNormalGenotypeCallSet}
\alias{getNormalGenotypeCallSet.TumorBoostNormalization}
\alias{getOutputDataSet.TumorBoostNormalization}
\alias{getOutputPath}
\alias{getOutputPath.TumorBoostNormalization}
\alias{getRootPath.TumorBoostNormalization}
\alias{nbrOfFiles.TumorBoostNormalization}
\alias{process.TumorBoostNormalization}
\alias{getAsteriskTags.TumorBoostNormalization}
\alias{getFullName.TumorBoostNormalization}
\alias{getName.TumorBoostNormalization}
\alias{getPath.TumorBoostNormalization}
\alias{getTags.TumorBoostNormalization}
\alias{setTags.TumorBoostNormalization}
\alias{getAsteriskTags.MultiSourceCopyNumberNormalization}
\alias{getTags.MultiSourceCopyNumberNormalization}
\alias{getFlavor}
\alias{getFlavor.TumorBoostNormalization}
\alias{as.character.AbstractCurveNormalization}
\alias{backtransformOne}
\alias{backtransformOne.AbstractCurveNormalization}
\alias{backtransformOne.PrincipalCurveNormalization}
\alias{backtransformOne.XYCurveNormalization}
\alias{fitOne.AbstractCurveNormalization}
\alias{fitOne.PrincipalCurveNormalization}
\alias{fitOne.XYCurveNormalization}
\alias{getAsteriskTags.AbstractCurveNormalization}
\alias{getDataSets.AbstractCurveNormalization}
\alias{getFullName.AbstractCurveNormalization}
\alias{getInputDataSet.AbstractCurveNormalization}
\alias{getName.AbstractCurveNormalization}
\alias{getOutputDataSet.AbstractCurveNormalization}
\alias{getPairedDataSet.AbstractCurveNormalization}
\alias{getPath.AbstractCurveNormalization}
\alias{getRootPath.AbstractCurveNormalization}
\alias{getSubsetToFit.AbstractCurveNormalization}
\alias{getTags.AbstractCurveNormalization}
\alias{getTargetDataSet.AbstractCurveNormalization}
\alias{nbrOfFiles.AbstractCurveNormalization}
\alias{process.AbstractCurveNormalization}
\alias{setTags.AbstractCurveNormalization}
\alias{getSubsetToFit.PrincipalCurveNormalization}
\alias{getPairedDataSet}
\alias{getTargetDataSet}
\alias{callGenotypes.RawAlleleBFractions}
\alias{getCalls}
\alias{getCalls.RawGenotypeCalls}
\alias{getColors}
\alias{getColors.RawGenotypeCalls}
\alias{isHeterozygous.RawGenotypeCalls}
\alias{isHomozygous.RawGenotypeCalls}
\alias{RawGenotypeCalls}
\alias{normalizeTumorBoost.RawAlleleBFractions}
\alias{plotSmoothedPairsOne}
\alias{plotSmoothedPairsOne.MultiSourceCopyNumberNormalization}
\alias{getBacktransforms}
\alias{getBacktransforms.PrincipalCurve}
\alias{getPrincipalCurveEstimator}
\alias{getPrincipalCurveEstimator.MultiSourceCopyNumberNormalization}
\alias{plotBacktransforms}
\alias{plotBacktransforms.PrincipalCurve}
\alias{getDataSet}
\alias{getDataSet.NaiveFracBGenotyping}
\alias{getGenotypeCallSet}
\alias{getGenotypeCallSet.NaiveFracBGenotyping}
\alias{NaiveFracBGenotyping}
\alias{nbrOfLoci.NaiveFracBGenotyping}
\alias{nbrOfSNPs}
\alias{nbrOfSNPs.NaiveFracBGenotyping}
\alias{process.NaiveFracBGenotyping}
\alias{processOne}
\alias{processOne.NaiveFracBGenotyping}
\alias{segmentByPruneCBS}
\alias{segmentByPruneCBS.RawGenomicSignals}
\alias{normalizeBAFsByRegions}
\alias{callAllelicBalanceByBAFs}
\alias{callCopyNeutralRegions}
\alias{getXScale.CopyNumberRegions}
\alias{getYScale.CopyNumberRegions}
\alias{plot.CopyNumberRegions}
\alias{segmentByPruneCBSv1}
\alias{segmentByPruneCBSv1.RawGenomicSignals}
\alias{orthogonalizeC1C2}
\alias{ABtoC1C2}
\alias{bootstrap}
\alias{bootstrap.PairedPSCBS}
\alias{C1C2toAB}
\alias{deShearC1C2}
\alias{extractC1C2}
\alias{extractC1C2.list}
\alias{fitLoess2D}
\alias{fitLoess2D.C1C2}
\alias{fitLoessKD}
\alias{fitLoessKD.data.frame}
\alias{fitLoessKD.matrix}
\alias{normalizeLoess2D}
\alias{normalizeLoess2D.C1C2}
\alias{normalizeLoessKD}
\alias{normalizeLoessKD.data.frame}
\alias{normalizeLoessKD.matrix}
\alias{plot.C1C2}
\alias{points.C1C2}
\alias{resampleA}
\alias{resampleA.PairedPSCBS}
\alias{resampleB}
\alias{resampleB.PairedPSCBS}
\alias{resampleC}
\alias{resampleC.PairedPSCBS}
\alias{callPeaks}
\alias{callPeaks.data.frame}
\alias{deShearC1C2_v0}
\alias{deShearC1C2_v0.PairedPSCBS}
\alias{extractDeltaC1C2}
\alias{extractDeltaC1C2.PairedPSCBS}
\alias{extractPolarDeltaC1C2}
\alias{extractPolarDeltaC1C2.PairedPSCBS}
\alias{fitC1C2ShearModel}
\alias{fitC1C2ShearModel.PairedPSCBS}
\alias{fitDeltaC1C2ShearModel}
\alias{fitDeltaC1C2ShearModel.PairedPSCBS}
\alias{fitDeltaXYShearModel}
\alias{fitDeltaXYShearModel.matrix}
\alias{calibrateC1C2}
\alias{estimateC2Bias}
\alias{estimateC2Bias.PairedPSCBS}
\alias{fitC1C2Densities}
\alias{fitC1C2Densities.PairedPSCBS}
\alias{fitC1C2Peaks}
\alias{fitC1C2Peaks.PairedPSCBS}
\alias{plotC1C2Grid}
\alias{plotC1C2Grid.PairedPSCBS}
\alias{translateC1C2}
\alias{translateC1C2.PairedPSCBS}
\alias{transformC1C2}
\alias{transformC1C2.PairedPSCBS}
\alias{PairedPSCBSFile}
\alias{PairedPSCBSFileSet}
\alias{byName.PairedPSCBSFileSet}
\alias{byPath.PairedPSCBSFileSet}
\alias{findAtomicAberrations.PairedPSCBS}
\alias{findByName.PairedPSCBSFileSet}
\alias{loadObject.PairedPSCBSFile}
\alias{plotTracks}
\alias{plotTracks.PruneCNA}
\alias{pruneCNA}
\alias{pruneCNA.PairedPSCBS}
\alias{testEqualityC1C2ByChiSquare}
\alias{testEqualityC1C2ByMean}
\alias{testEqualityTcnByMean}
\alias{testEqualityTcnByT}
\alias{cat}
\alias{getOption}
\alias{drawC1C2Density}
\alias{drawC1C2Density.PairedPSCBS}
\alias{getAsteriskTags.TotalCnBinnedSmoothing}
\alias{getParameters.TotalCnBinnedSmoothing}
\alias{smoothRawCopyNumbers.TotalCnBinnedSmoothing}
\alias{findAtomicAberrations.CBS}
\alias{pruneCNA.AbstractCBS}
\alias{[.PruneCNA}
\alias{extractGenerations}
\alias{extractGenerations.PruneCNA}
\alias{nbrOfGenerations}
\alias{nbrOfGenerations.PruneCNA}
\alias{testEqualityC1C2ByChiSq}
\alias{testEqualityC1orC2ByT}
\alias{as.character.PairedPscbsModel}
\alias{fit.PairedPscbsModel}
\alias{getAsteriskTags.PairedPscbsModel}
\alias{getChipType.PairedPscbsModel}
\alias{getChromosomes.PairedPscbsModel}
\alias{getDataSets.PairedPscbsModel}
\alias{getFitFunction.PairedPscbsModel}
\alias{getFullName.PairedPscbsModel}
\alias{getName.PairedPscbsModel}
\alias{getNormalDataSet.PairedPscbsModel}
\alias{getOptionalArguments.PairedPscbsModel}
\alias{getParameters.PairedPscbsModel}
\alias{getParametersAsString.PairedPscbsModel}
\alias{getPath.PairedPscbsModel}
\alias{getRandomSeed.PairedPscbsModel}
\alias{getRootPath.PairedPscbsModel}
\alias{getTags.PairedPscbsModel}
\alias{getTumorDataSet}
\alias{getTumorDataSet.PairedPscbsModel}
\alias{indexOf.PairedPscbsModel}
\alias{nbrOfFiles.PairedPscbsModel}
\alias{setRandomSeed.PairedPscbsModel}
\alias{setTags.PairedPscbsModel}
\alias{calculatePairedPSCNByGenotype}
\alias{calculatePairedPSCNByGenotype.data.frame}
\alias{calculatePairedPSCNByGenotype.numeric}
\alias{calculateTumorPSCNByGenotypeUsingCTCN}
\alias{calculateTumorPSCNByGenotypeUsingCTCN.numeric}
\alias{PairedPscbsCaller}
\alias{getAsteriskTags.PairedPscbsCaller}
\alias{getParameters.PairedPscbsCaller}
\alias{getPath.PairedPscbsCaller}
\alias{getPlatform.PairedPscbsCaller}
\alias{getRootPath.PairedPscbsCaller}
\alias{process.PairedPscbsCaller}
\alias{getOutputDataSet.PairedPscbsModel}
\alias{loadFit.PairedPSCBSFile}
\alias{getChipType.PairedPSCBSFileSet}
\alias{getDefaultFullName.PairedPSCBSFileSet}
\alias{getPlatform.PairedPSCBSFileSet}
\alias{getOptionalArguments.PairedPscbsCaller}
\alias{backgroundCorrect}
\alias{backgroundCorrect.PairedPSCBS}
\alias{getOutputDataSet0.TotalCnSmoothing}
\alias{getOutputFileClass}
\alias{getOutputFileClass.TotalCnSmoothing}
\alias{getOutputFileExtension}
\alias{getOutputFileExtension.TotalCnSmoothing}
\alias{getOutputFileSetClass}
\alias{getOutputFileSetClass.TotalCnSmoothing}
\alias{adjustAB}
\alias{adjustAB.PSCBS}
\alias{callChangePointFlips}
\alias{callChangePointFlips.PSCBS}
\alias{callChangePointFlips.data.frame}
\alias{callChangePointFlips.matrix}
\alias{deshearC1C2}
\alias{deshearC1C2.PSCBS}
\alias{estimateChangePointFlips}
\alias{estimateChangePointFlips.data.frame}
\alias{estimateChangePointFlips.matrix}
\alias{flipChangePoints}
\alias{flipChangePoints.PSCBS}
\alias{estimateDeshearingParameter}
\alias{estimateDeshearingParameter.PSCBS}
\alias{deShearC1C2_20120922}
\alias{estimateDeshearingParameter_20130815}
\alias{estimateDeshearingParameter_20130815.PSCBS}

\description{
  This page contains aliases for all "non-documented" objects that
  \code{R CMD check} detects in this package.

  Almost all of them are \emph{generic} functions that have specific
  document for the corresponding method coupled to a specific class.
  Other functions are re-defined by \code{setMethodS3()} to
  \emph{default} methods. Neither of these two classes are non-documented
  in reality.
  The rest are deprecated methods.
}


\keyword{documentation}
\keyword{internal}
